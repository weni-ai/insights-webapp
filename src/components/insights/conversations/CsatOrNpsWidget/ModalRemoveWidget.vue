<template>
  <UnnnicModalDialog
    data-testid="modal-remove-widget"
    class="modal-remove-widget"
    :modelValue="modelValue"
    :title="
      $tc(
        'conversations_dashboard.customize_your_dashboard.modal_remove_widget.title',
        { type: type.toUpperCase() },
      )
    "
    type="warning"
    icon="warning"
    :primaryButtonProps="{
      text: $t(
        'conversations_dashboard.customize_your_dashboard.modal_remove_widget.remove',
      ),
      onClick: handleRemoveWidget,
    }"
    :secondaryButtonProps="{
      text: $t(
        'conversations_dashboard.customize_your_dashboard.modal_remove_widget.cancel',
      ),
      onClick: () => emit('update:modelValue', false),
    }"
    showCloseIcon
    size="sm"
    @update:model-value="emit('update:modelValue', $event)"
  >
    <p class="modal-remove-widget__description">
      {{
        $tc(
          'conversations_dashboard.customize_your_dashboard.modal_remove_widget.description',
          { type: type.toUpperCase() },
        )
      }}
    </p>
  </UnnnicModalDialog>
</template>

<script setup lang="ts">
import { useConversationalWidgets } from '@/store/modules/conversational/widgets';

interface Props {
  type: 'csat' | 'nps';
  modelValue: boolean;
}

const props = defineProps<Props>();

const emit = defineEmits<{
  (_e: 'update:modelValue', _value: boolean): void;
}>();

const { deleteWidget } = useConversationalWidgets();

const handleRemoveWidget = () => {
  deleteWidget(props.type);
  emit('update:modelValue', false);
};
</script>

<style lang="scss" scoped>
.modal-remove-widget {
  display: flex;
  flex-direction: column;
  gap: $unnnic-spacing-md;

  &__description {
    font-size: $unnnic-font-size-body-gt;
    color: $unnnic-color-neutral-cloudy;
  }

  :deep(.unnnic-modal-dialog__container__content) {
    padding-bottom: $unnnic-spacing-sm;
  }

  :deep(.unnnic-modal-dialog__container__actions) {
    padding-top: $unnnic-spacing-sm;
  }
}
</style>
