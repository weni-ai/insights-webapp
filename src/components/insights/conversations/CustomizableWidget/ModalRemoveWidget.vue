<template>
  <UnnnicModalDialog
    data-testid="modal-remove-widget"
    class="modal-remove-widget"
    :modelValue="modelValue"
    :title="
      $t(
        'conversations_dashboard.customize_your_dashboard.modal_remove_widget.title',
      )
    "
    type="warning"
    icon="warning"
    :primaryButtonProps="{
      text: $t(
        'conversations_dashboard.customize_your_dashboard.modal_remove_widget.remove',
      ),
      loading: isLoading,
      onClick: handleRemoveWidget,
    }"
    :secondaryButtonProps="{
      text: $t(
        'conversations_dashboard.customize_your_dashboard.modal_remove_widget.cancel',
      ),
      onClick: () => emit('update:modelValue', false),
    }"
    showCloseIcon
    size="sm"
    @update:model-value="emit('update:modelValue', $event)"
  >
    <p
      class="modal-remove-widget__description"
      data-testid="modal-remove-widget-description"
    >
      {{
        $t(
          'conversations_dashboard.customize_your_dashboard.modal_remove_widget.description',
          {
            type: props.name,
          },
        )
      }}
    </p>
  </UnnnicModalDialog>
</template>

<script setup lang="ts">
import { ref } from 'vue';
import { useCustomWidgets } from '@/store/modules/conversational/customWidgets';
import { useConversationalWidgets } from '@/store/modules/conversational/widgets';
import { unnnicCallAlert } from '@weni/unnnic-system';
import { useI18n } from 'vue-i18n';

const { t } = useI18n();

interface Props {
  type: 'csat' | 'nps' | 'custom' | 'sales_funnel';
  modelValue: boolean;
  uuid?: string;
  name?: string;
}

const props = defineProps<Props>();

const emit = defineEmits<{
  (_e: 'update:modelValue', _value: boolean): void;
}>();

const { deleteWidget } = useConversationalWidgets();
const { deleteCustomWidget } = useCustomWidgets();

const isLoading = ref(false);

const handleRemoveWidget = async () => {
  try {
    isLoading.value = true;
    if (props.type === 'custom') {
      await deleteCustomWidget(props.uuid);
    } else {
      await deleteWidget(props.type);
    }

    unnnicCallAlert({
      props: {
        version: '1.1',
        text: t(
          'conversations_dashboard.customize_your_dashboard.modal_remove_widget.success_message',
          {
            widget: props.name,
          },
        ),
        type: 'success',
        seconds: 5,
      },
    });

    emit('update:modelValue', false);
  } catch (error) {
    console.error(error);
  } finally {
    isLoading.value = false;
  }
};
</script>

<style lang="scss" scoped>
.modal-remove-widget {
  display: flex;
  flex-direction: column;
  gap: $unnnic-spacing-md;

  &__description {
    font-size: $unnnic-font-size-body-gt;
    color: $unnnic-color-neutral-cloudy;
  }

  :deep(.unnnic-modal-dialog__container__content) {
    padding-bottom: $unnnic-spacing-sm;
  }

  :deep(.unnnic-modal-dialog__container__actions) {
    padding-top: $unnnic-spacing-sm;
  }
}
</style>
