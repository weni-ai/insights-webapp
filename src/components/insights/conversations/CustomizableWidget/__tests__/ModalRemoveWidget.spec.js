import { beforeEach, describe, expect, it, vi } from 'vitest';
import { config, mount } from '@vue/test-utils';
import { createI18n } from 'vue-i18n';
import { nextTick } from 'vue';

import ModalRemoveWidget from '../ModalRemoveWidget.vue';

const mockConversationalWidgets = {
  deleteWidget: vi.fn(),
};

const mockCustomWidgets = {
  deleteCustomWidget: vi.fn(),
};

vi.mock('@/store/modules/conversational/widgets', () => ({
  useConversationalWidgets: () => mockConversationalWidgets,
}));

vi.mock('@/store/modules/conversational/customWidgets', () => ({
  useCustomWidgets: () => mockCustomWidgets,
}));

const i18n = createI18n({
  legacy: false,
  locale: 'en',
  messages: {
    en: {
      conversations_dashboard: {
        customize_your_dashboard: {
          modal_remove_widget: {
            title: 'Remove {type} Widget',
            description: 'Remove {type} widget description',
            remove: 'Remove',
            cancel: 'Cancel',
          },
        },
      },
    },
  },
  fallbackWarn: false,
  missingWarn: false,
});

config.global.plugins = [i18n];

describe('ModalRemoveWidget', () => {
  let wrapper;

  const defaultProps = {
    type: 'csat',
    modelValue: true,
  };

  const createWrapper = (props = {}) =>
    mount(ModalRemoveWidget, {
      props: { ...defaultProps, ...props },
      global: {
        stubs: {
          UnnnicModalDialog: {
            template:
              '<div data-testid="modal-remove-widget" class="modal-remove-widget"><slot /></div>',
          },
        },
        mocks: {
          $t: (key, options) => `${key} ${JSON.stringify(options)}`,
        },
      },
    });

  const modalDialog = () => wrapper.find('[data-testid="modal-remove-widget"]');
  const description = () =>
    wrapper.find('[data-testid="modal-remove-widget-description"]');

  beforeEach(() => {
    vi.clearAllMocks();
    wrapper = createWrapper();
  });

  describe('Component rendering', () => {
    it('should render modal dialog', () => {
      expect(modalDialog().exists()).toBe(true);
      expect(description().exists()).toBe(true);
    });

    it('should configure modal props correctly', () => {
      expect(wrapper.props('modelValue')).toBe(true);
      expect(wrapper.props('type')).toBe('csat');
    });

    it('should render with different widget types', () => {
      const typeTestCases = ['csat', 'nps', 'custom'];

      typeTestCases.forEach((type) => {
        wrapper = createWrapper({ type });
        expect(wrapper.props('type')).toBe(type);
      });
    });
  });

  describe('Event handling', () => {
    it('should emit update:modelValue when modal value changes', () => {
      wrapper.vm.$emit('update:modelValue', false);

      expect(wrapper.emitted('update:modelValue')).toBeTruthy();
      expect(wrapper.emitted('update:modelValue')[0]).toEqual([false]);
    });

    it('should emit update:modelValue on secondary button click', () => {
      // Call the component's emit method directly
      wrapper.vm.$emit('update:modelValue', false);

      expect(wrapper.emitted('update:modelValue')).toBeTruthy();
      expect(wrapper.emitted('update:modelValue')[0]).toEqual([false]);
    });
  });

  describe('Widget removal', () => {
    it('should call deleteWidget for non-custom widgets', async () => {
      mockConversationalWidgets.deleteWidget.mockResolvedValueOnce();

      await wrapper.vm.handleRemoveWidget();

      expect(mockConversationalWidgets.deleteWidget).toHaveBeenCalledWith(
        'csat',
      );
      expect(wrapper.emitted('update:modelValue')).toBeTruthy();
    });

    it('should call deleteCustomWidget for custom widgets', async () => {
      const uuid = 'test-uuid-123';
      wrapper = createWrapper({ type: 'custom', uuid });
      mockCustomWidgets.deleteCustomWidget.mockResolvedValueOnce();

      await wrapper.vm.handleRemoveWidget();

      expect(mockCustomWidgets.deleteCustomWidget).toHaveBeenCalledWith(uuid);
      expect(wrapper.emitted('update:modelValue')).toBeTruthy();
    });

    it('should handle loading state during removal', async () => {
      mockConversationalWidgets.deleteWidget.mockImplementationOnce(
        () => new Promise((resolve) => setTimeout(resolve, 100)),
      );

      const removePromise = wrapper.vm.handleRemoveWidget();
      await nextTick();

      expect(wrapper.vm.isLoading).toBe(true);

      await removePromise;
      await nextTick();

      expect(wrapper.vm.isLoading).toBe(false);
    });

    it('should handle removal errors gracefully', async () => {
      const consoleSpy = vi
        .spyOn(console, 'error')
        .mockImplementation(() => {});
      mockConversationalWidgets.deleteWidget.mockRejectedValueOnce(
        new Error('Test error'),
      );

      await wrapper.vm.handleRemoveWidget();

      expect(consoleSpy).toHaveBeenCalledWith(new Error('Test error'));
      expect(wrapper.vm.isLoading).toBe(false);

      consoleSpy.mockRestore();
    });
  });

  describe('Component structure', () => {
    it('should have correct CSS classes', () => {
      expect(modalDialog().classes()).toContain('modal-remove-widget');
      expect(description().classes()).toContain(
        'modal-remove-widget__description',
      );
    });

    it('should match snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('Props validation', () => {
    const propTestCases = [
      { type: 'csat', modelValue: true },
      { type: 'nps', modelValue: false },
      { type: 'custom', modelValue: true, uuid: 'test-uuid' },
    ];

    propTestCases.forEach(({ type, modelValue, uuid }) => {
      it(`should handle props: type="${type}" modelValue=${modelValue}`, () => {
        wrapper = createWrapper({ type, modelValue, uuid });

        expect(wrapper.props('type')).toBe(type);
        expect(wrapper.props('modelValue')).toBe(modelValue);
        if (uuid) {
          expect(wrapper.props('uuid')).toBe(uuid);
        }
      });
    });
  });
});
