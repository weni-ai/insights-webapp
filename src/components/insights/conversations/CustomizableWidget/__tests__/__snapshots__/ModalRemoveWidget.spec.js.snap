// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ModalRemoveWidget > Component structure > should match snapshot 1`] = `
<div
  class="modal-remove-widget"
  data-testid="modal-remove-widget"
  data-v-1d7c4579=""
  icon="warning"
  modelvalue="true"
  primarybuttonprops="[object Object]"
  secondarybuttonprops="[object Object]"
  showcloseicon=""
  size="sm"
  title="conversations_dashboard.customize_your_dashboard.modal_remove_widget.title {"type":"CSAT"}"
  type="warning"
>
  
  <p
    class="modal-remove-widget__description"
    data-testid="modal-remove-widget-description"
    data-v-1d7c4579=""
  >
    conversations_dashboard.customize_your_dashboard.modal_remove_widget.description {"type":"CSAT"}
  </p>
  
</div>
`;
